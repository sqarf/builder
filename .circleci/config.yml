# Visual Composer Website Builder circleci config file

version: 2
jobs:
  js-codestyle:
    docker:
      - image: visualcomposer/ci-wordpress:php7.3
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-node-codestyle-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
            - v1-node-codestyle-{{ arch }}-{{ .Branch }}-
            - v1-node-codestyle-{{ arch }}-
      - run: npm install -g n
      - run: n 14
      - run: yarn install --frozen-lockfile && yarn standard
      - save_cache:
          key: v1-node-codestyle-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
          paths:
            - ./node_modules
  php-codestyle:
    docker:
      - image: visualcomposer/ci-wordpress:php7.3
    steps:
      - checkout
      - run:
          name: CodeStyle check
          command: php ci/phpcs.phar --standard=ci/ruleset.xml plugin-wordpress.php bootstrap visualcomposer/Framework visualcomposer/Helpers visualcomposer/Modules visualcomposer/Application.php visualcomposer/Requirements.php
      - run:
          name: PHPGrep check
          command: ci/phpgrep-lint.sh visualcomposer/ ci/patterns.txt
  php-tests:
    docker:
      - image: visualcomposer/ci-wordpress:circleci-php-apache-chrome
        environment:
          MYSQL_HOST: 127.0.0.1
          MYSQL_DB: mate
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
          WP_TESTS_DIR: /opt/wp-tests
          WP_TESTS_DOMAIN: localhost
          WP_TESTS_ABSPATH: /var/www/html/
          WP_TESTS_DB_NAME: test
          WP_TESTS_DB_USER: wpuser
          WP_TESTS_DB_HOST: 127.0.0.1
      - image: mariadb:10.3.13-bionic
        command: mysqld --character-set-server=utf8mb4 --collation-server=utf8mb4_bin --innodb-large-prefix=true --innodb-file-format=Barracuda
        environment:
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
    steps:
      - checkout
      - run:
          name: Wait for db
          command: dockerize -wait tcp://localhost:3306 -timeout 10s
      - run:
          name: Setup mysql and apache2
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="CREATE DATABASE test;"
            mysql -h 127.0.0.1 --user="root" --execute="CREATE USER 'wpuser'@'127.0.0.1'; GRANT ALL PRIVILEGES ON test.* TO 'wpuser'@'127.0.0.1';"
            sudo apachectl start
      - run:
          name: Install Wordpress
          command: |
            wp core config --path=/var/www/html --dbhost=127.0.0.1 --dbname=test --dbuser=wpuser --allow-root
            wp core install --url=http://localhost --title=Test --admin_user=admin --admin_password=12345 --admin_email=test@test.com --path=/var/www/html --skip-email --allow-root --debug
            sudo chmod 777 -R /var/www/html/wp-content
      - run: php ci/composer.phar install
      - run: php tools/php-composer/cli.php
      - run: php -d 'zend_extension=xdebug.so' ci/phpunit-9.0.0.phar --configuration=ci/phpunit-all-coverage.xml
      - run: php ci/coverage-test.php
  end-2-end-tests-chrome:
    docker:
      - image: visualcomposer/ci-wordpress:circleci-php-apache-chrome
        environment:
          MYSQL_HOST: 127.0.0.1
          MYSQL_DB: mate
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
          WP_TESTS_DIR: /opt/wp-tests
          WP_TESTS_DOMAIN: localhost
          WP_TESTS_ABSPATH: /var/www/html/
      - image: mariadb:10.3.13-bionic
        command: mysqld --character-set-server=utf8mb4 --collation-server=utf8mb4_bin --innodb-large-prefix=true --innodb-file-format=Barracuda
        environment:
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
    steps:
      - checkout
      - run:
          name: Wait for db
          command: dockerize -wait tcp://localhost:3306 -timeout 1m
      - run:
          name: Setup mysql and apache2
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="CREATE DATABASE test;"
            mysql -h 127.0.0.1 --user="root" --execute="CREATE USER 'wpuser'@'127.0.0.1'; GRANT ALL PRIVILEGES ON test.* TO 'wpuser'@'127.0.0.1';"
            sudo apachectl start
      - run:
          name: Install Wordpress
          command: |
            wp core config --path=/var/www/html --dbhost=127.0.0.1 --dbname=test --dbuser=wpuser --allow-root
            wp core install --url=http://localhost --title=Test --admin_user=admin --admin_password=12345 --admin_email=test@test.com --path=/var/www/html --skip-email --allow-root --debug
      - run: php ci/composer.phar install
      - run:
          name: Link files and directories
          command: |
            echo "<?php VcvEnv::set('VCV_ENV_DEV_ELEMENTS', true); VcvEnv::set('VCV_ENV_DEV_ADDONS', true); VcvEnv::set('VCV_ENV_EXTENSION_DOWNLOAD', false); require_once(__DIR__ . '/env.php');" > env-dev.php
            ln -s $(pwd) /var/www/html/wp-content/plugins/builder
            sudo chmod 777 -R /var/www/html/wp-content
      - restore_cache:
          keys:
            - v1.1-node-e2e-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
            - v1.1-node-e2e-{{ arch }}-{{ .Branch }}-
            - v1.1-node-e2e-{{ arch }}-
      - run:
          name: Install node modules
          command: yarn install --frozen-lockfile
      - save_cache:
          name: Cache node_modules
          key: v1.1-node-e2e-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
          paths:
            - ./node_modules
      - run:
          name: Build project
          command: |
            yarn build
            yarn build-settings
            bash tools/elements/buildScript.sh
      - run:
          name: Activate plugin
          command: wp plugin activate --all --path=/var/www/html
      - run:
          name: Add free license to db
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-key', '0b147957-48d2-4c15-b50e-f9f755bd8749', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-type', 'free', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-expiration', 'lifetime', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-settings-initial-helpers-enabled', '', 'yes');"
      - run: yarn run cypress install
      - run: yarn run cypress run --browser=chrome --project ./tests/cypressChrome
      - store_artifacts:
          path: ./tests/cypressChrome/cypress/snapshots
      - store_artifacts:
          path: ./tests/cypressChrome/cypress/videos
      - store_artifacts:
          path: ./tests/cypressChrome/cypress/screenshots
  end-2-end-tests-electron:
    docker:
      - image: visualcomposer/ci-wordpress:circleci-php-apache-chrome
        environment:
          MYSQL_HOST: 127.0.0.1
          MYSQL_DB: mate
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
          WP_TESTS_DIR: /opt/wp-tests
          WP_TESTS_DOMAIN: localhost
          WP_TESTS_ABSPATH: /var/www/html/
      - image: mariadb:10.3.13-bionic
        command: mysqld --character-set-server=utf8mb4 --collation-server=utf8mb4_bin --innodb-large-prefix=true --innodb-file-format=Barracuda
        environment:
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
    steps:
      - checkout
      - run:
          name: Wait for db
          command: dockerize -wait tcp://localhost:3306 -timeout 1m
      - run:
          name: Setup mysql and apache2
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="CREATE DATABASE test;"
            mysql -h 127.0.0.1 --user="root" --execute="CREATE USER 'wpuser'@'127.0.0.1'; GRANT ALL PRIVILEGES ON test.* TO 'wpuser'@'127.0.0.1';"
            sudo apachectl start
      - run:
          name: Install Wordpress
          command: |
            wp core config --path=/var/www/html --dbhost=127.0.0.1 --dbname=test --dbuser=wpuser --allow-root
            wp core install --url=http://localhost --title=Test --admin_user=admin --admin_password=12345 --admin_email=test@test.com --path=/var/www/html --skip-email --allow-root --debug
      - run: php ci/composer.phar install
      - run:
          name: Link files and directories
          command: |
            echo "<?php VcvEnv::set('VCV_ENV_DEV_ELEMENTS', false); VcvEnv::set('VCV_ENV_DEV_ADDONS', false); VcvEnv::set('VCV_ENV_EXTENSION_DOWNLOAD', true); require_once(__DIR__ . '/env.php');" > env-dev.php
            ln -s $(pwd) /var/www/html/wp-content/plugins/builder
            sudo chmod 777 -R /var/www/html/wp-content
      - restore_cache:
          keys:
            - v1.1-node-e2e-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
            - v1.1-node-e2e-{{ arch }}-{{ .Branch }}-
            - v1.1-node-e2e-{{ arch }}-
      - run:
          name: Install node modules
          command: yarn install --frozen-lockfile
      - save_cache:
          name: Cache node_modules
          key: v1.1-node-e2e-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
          paths:
            - ./node_modules
      - run:
          name: Build project
          command: |
            yarn build
            yarn build-settings
            bash tools/elements/buildScript.sh
      - run:
          name: Activate plugin
          command: wp plugin activate --all --path=/var/www/html
      - run:
          name: Add free license to db
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-key', '0b147957-48d2-4c15-b50e-f9f755bd8749', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-type', 'free', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-expiration', 'lifetime', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-settings-initial-helpers-enabled', '', 'yes');"
      - run: yarn run cypress install
      - run: yarn run cypress run --project ./tests/cypressElectron
      - store_artifacts:
          path: ./tests/cypressElectron/cypress/snapshots
      - store_artifacts:
          path: ./tests/cypressElectron/cypress/videos
      - store_artifacts:
          path: ./tests/cypressElectron/cypress/screenshots
  end-2-end-tests-pipeline:
    docker:
      - image: visualcomposer/ci-wordpress:circleci-php-apache-chrome
        environment:
          MYSQL_HOST: 127.0.0.1
          MYSQL_DB: mate
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
          WP_TESTS_DIR: /opt/wp-tests
          WP_TESTS_DOMAIN: localhost
          WP_TESTS_ABSPATH: /var/www/html/
      - image: mariadb:10.3.13-bionic
        command: mysqld --character-set-server=utf8mb4 --collation-server=utf8mb4_bin --innodb-large-prefix=true --innodb-file-format=Barracuda
        environment:
          MYSQL_USER: root
          MYSQL_ALLOW_EMPTY_PASSWORD: true
    steps:
      - checkout
      - run:
          name: Wait for db
          command: dockerize -wait tcp://localhost:3306 -timeout 1m
      - run:
          name: Setup mysql and apache2
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="CREATE DATABASE test;"
            mysql -h 127.0.0.1 --user="root" --execute="CREATE USER 'wpuser'@'127.0.0.1'; GRANT ALL PRIVILEGES ON test.* TO 'wpuser'@'127.0.0.1';"
            sudo apachectl start
      - run:
          name: Install Wordpress
          command: |
            wp core config --path=/var/www/html --dbhost=127.0.0.1 --dbname=test --dbuser=wpuser --allow-root
            wp core install --url=http://localhost --title=Test --admin_user=admin --admin_password=12345 --admin_email=test@test.com --path=/var/www/html --skip-email --allow-root --debug
      - run: php ci/composer.phar install
      - run:
          name: Link files and directories
          command: |
            echo "<?php VcvEnv::set('VCV_ENV_DEV_ELEMENTS', false); VcvEnv::set('VCV_ENV_DEV_ADDONS', false); VcvEnv::set('VCV_ENV_EXTENSION_DOWNLOAD', true); require_once(__DIR__ . '/env.php');" > env-dev.php
            ln -s $(pwd) /var/www/html/wp-content/plugins/builder
            sudo chmod 777 -R /var/www/html/wp-content
      - restore_cache:
          keys:
            - v1.1-node-e2e-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
            - v1.1-node-e2e-{{ arch }}-{{ .Branch }}-
            - v1.1-node-e2e-{{ arch }}-
      - run:
          name: Install node modules
          command: yarn install --frozen-lockfile
      - save_cache:
          name: Cache node_modules
          key: v1.1-node-e2e-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
          paths:
            - ./node_modules
      - run:
          name: Build project
          command: |
            yarn build
            yarn build-settings
            bash tools/elements/buildScript.sh
      - run:
          name: Activate plugin
          command: wp plugin activate --all --path=/var/www/html
      - run:
          name: Add free license to db
          command: |
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-key', '0b147957-48d2-4c15-b50e-f9f755bd8749', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-type', 'free', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-license-expiration', 'lifetime', 'yes');"
            mysql -h 127.0.0.1 --user="root" --execute="REPLACE INTO test.wp_options (option_name, option_value, autoload)  VALUES ('vcv-settings-initial-helpers-enabled', '', 'yes');"
      - run: yarn run cypress install
      - run: yarn run cypress run --project ./tests/cypressPipeline
      - store_artifacts:
          path: ./tests/cypressPipeline/cypress/snapshots
      - store_artifacts:
          path: ./tests/cypressPipeline/cypress/videos
      - store_artifacts:
          path: ./tests/cypressPipeline/cypress/screenshots
  js-stores-tests:
    docker:
      - image: visualcomposer/ci-wordpress:php7.3
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-node-js-stores-tests-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
            - v1-node-js-stores-tests-{{ arch }}-{{ .Branch }}-
            - v1-node-js-stores-tests-{{ arch }}-
      - run: npm install -g n
      - run: n 14
      - run: yarn install
      - run: yarn build
      - run: yarn test -u
      - save_cache:
          key: v1-node-js-stores-tests-{{ arch }}-{{ .Branch }}-{{ checksum "package.json" }}
          paths:
            - ./node_modules
workflows:
  version: 2
  codestyle:
    jobs:
      - js-codestyle
      - php-codestyle
  tests:
    jobs:
      - php-tests
      - js-stores-tests
      - end-2-end-tests-pipeline
