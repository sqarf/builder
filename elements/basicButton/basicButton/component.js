import React from 'react'
import vcCake from 'vc-cake'

const vcvAPI = vcCake.getService('api')

export default class ButtonElement extends vcvAPI.elementComponent {
  getColorSelector (textColor, backgroundColor) {
    const txtColor = [...textColor.matchAll(/[\da-f]+/gi)].map(match => match[0]).join('-')
    const bgColor = [...backgroundColor.matchAll(/[\da-f]+/gi)].map(match => match[0]).join('-')
    return `${bgColor}--${txtColor}`
  }

  render () {
    const { id, atts, editor } = this.props
    const { buttonUrl, buttonText, shape, alignment, customClass, toggleCustomHover, metaCustomId, size, toggleStretchButton, color, background, hoverColor, hoverBackground } = atts

    let containerClasses = 'vce-button--style-basic-container'
    let wrapperClasses = 'vce-button--style-basic-wrapper vce'
    let classes = 'vce-button vce-button--style-basic'
    let customProps = {}
    let CustomTag = 'button'

    if (buttonUrl && buttonUrl.url) {
      CustomTag = 'a'
      const { url, title, targetBlank, relNofollow } = buttonUrl
      customProps = {
        href: url,
        title: title,
        target: targetBlank ? '_blank' : undefined,
        rel: relNofollow ? 'nofollow' : undefined
      }
    }

    if (typeof customClass === 'string' && customClass) {
      containerClasses += ' ' + customClass
    }

    if (shape) {
      classes += ` vce-button--style-basic--border-${shape}`
    }

    if (alignment) {
      containerClasses += ` vce-button--style-basic-container--align-${alignment}`
    }

    if (size) {
      classes += ` vce-button--style-basic--size-${size}`
    }

    if (toggleStretchButton) {
      wrapperClasses += ' vce-button--style-basic-wrapper--stretched'
    }

    classes += ` vce-button--style-basic--color-${this.getColorSelector(color, background)}`

    if (toggleCustomHover) {
      classes += ` vce-button--style-basic--hover-color-${this.getColorSelector(hoverColor, hoverBackground)}`
    }

    if (metaCustomId) {
      customProps.id = metaCustomId
    }

    const doMargin = this.applyDO('margin')
    const doRest = this.applyDO('padding border background animation')

    return (
      <div className={containerClasses} {...editor}>
        <span className={wrapperClasses} id={'el-' + id} {...doMargin}>
          <CustomTag className={classes} {...customProps} {...doRest}>
            {buttonText}
          </CustomTag>
        </span>
      </div>
    )
  }
}
