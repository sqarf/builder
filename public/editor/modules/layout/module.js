/* global VCV_PLUGIN_UPDATE */
import vcCake from 'vc-cake'
import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import store from 'public/editor/stores/store'
import { notificationAdded } from 'public/editor/stores/notifications/slice'
import Editor from './lib/editor'
import DndManager from './lib/dndManager'
import ControlsManager from './lib/controlsIframe/controlsManager'
import RightClickMenu from 'public/components/rightClickMenu/rightClickMenu'
import MobileControlsManager from './lib/controlsIframe/mobileControlsManager'
import MobileDetect from 'mobile-detect'
import OopsScreen from 'public/components/account/oopsScreen'
import Notifications from 'public/components/notifications/notifications'
import Popup from 'public/components/popup/popupContainer'
import FullPagePopupContainer from 'public/components/popup/fullPagePopupContainer'
import Helpers from 'public/components/helpers/helpers'

const Utils = vcCake.getService('utils')
const workspaceStorage = vcCake.getStorage('workspace')
const wordpressDataStorage = vcCake.getStorage('wordpressData')
const workspaceSettings = workspaceStorage.state('settings')
const workspaceIFrame = workspaceStorage.state('iframe')
const elementsStorage = vcCake.getStorage('elements')
const assetsStorage = vcCake.getStorage('assets')
const dataManager = vcCake.getService('dataManager')

vcCake.add('contentLayout', (api) => {
  // disable this module on post update action
  if (typeof window.vcvRebuildPostSave !== 'undefined') {
    return
  }
  const iframeContent = document.getElementById('vcv-layout-iframe-content')
  const dnd = new DndManager(api)
  const controls = new ControlsManager(api)
  const rightClick = new RightClickMenu()

  const localizations = dataManager.get('localizations')
  if (Utils.isRTL()) {
    document.body && document.body.classList.add('rtl')
  }

  // Start notifications and editor popups
  const layoutOverlay = document.querySelector('.vcv-layout-overlay')
  if (layoutOverlay) {
    ReactDOM.render(
      <Provider store={store}>
        <Notifications />
        <Popup />
        <FullPagePopupContainer />
        {dataManager.get('showInitialHelpers') && <Helpers />}
      </Provider>,
      layoutOverlay
    )
  }

  const renderLayout = (reload = false) => {
    workspaceIFrame.ignoreChange(reloadLayout)
    workspaceIFrame.set(false)
    const iframe = document.getElementById('vcv-editor-iframe')
    const iframeWindow = iframe ? iframe.contentWindow : null
    const domContainer = iframeWindow ? iframeWindow.document.getElementById('vcv-editor') : null
    if (domContainer) {
      if (iframeWindow.vcv) {
        iframeWindow.vcv.trigger('ready')
      }
      ReactDOM.render(
        <Provider store={store}>
          <Editor api={api} />
        </Provider>,
        domContainer
      )

      !reload && dnd.init()
      rightClick.init()

      workspaceIFrame.onChange(reloadLayout)

      const updateNotificationEnabled = vcCake.env('VCV_FT_UPDATE_NOTIFICATION')
      const pluginUpdate = typeof VCV_PLUGIN_UPDATE === 'function' ? VCV_PLUGIN_UPDATE() : false
      pluginUpdate && updateNotificationEnabled && store.dispatch(notificationAdded({
        showCloseButton: true,
        type: 'warning',
        text: localizations.newPluginVersionIsAvailable || 'There is a new version of Visual Composer Website Builder available',
        html: true,
        time: -1
      }))

      const mobileDetect = new MobileDetect(window.navigator.userAgent)
      if (mobileDetect.mobile() && (mobileDetect.tablet() || mobileDetect.phone())) {
        const mobileControls = new MobileControlsManager(api)
        mobileControls.init()
        store.dispatch(notificationAdded({
          text: localizations.mobileTooltipText || 'Double-tap on an element to open the edit window. Tap and hold to initiate drag and drop.',
          cookie: 'vcv-mobile-tooltip',
          time: 10000
        }))
        return
      }

      reload ? controls.updateIframeVariables() : controls.init()
      if (vcCake.env('VCV_JS_THEME_LAYOUTS')) {
        iframeWindow.document.querySelectorAll('[data-vcv-layout-zone]').forEach((zone) => {
          const zoneButton = zone.querySelector('[data-vcv-action="settings"]')
          zoneButton && zoneButton.addEventListener('click', () => {
            workspaceStorage.state('content').set('settings')
            workspaceSettings.set({ action: 'settings' })
          })
        })
      }
    } else {
      const postUpdateWrapper = document.querySelector('#vcv-posts-update-wrapper')
      let wrapper = document.body

      if (postUpdateWrapper) {
        const postUpdateIframe = postUpdateWrapper.querySelector('.vcv-layout-iframe')
        wrapper = postUpdateIframe && postUpdateIframe.contentWindow.document.body
      }

      wrapper.innerHTML = '<div id=\'vcv-oops-screen-container\'></div>'
      const oopsContainer = document.getElementById('vcv-oops-screen-container')
      if (oopsContainer) {
        ReactDOM.render(
          <div className='vcv-screen-section'>
            <OopsScreen errorName={dataManager.get('frontEndError')} />
          </div>,
          oopsContainer
        )
      }
    }
  }

  const removeLoadingScreen = () => {
    let loadingOverlays = iframeContent.querySelectorAll('.vcv-loading-overlay')
    loadingOverlays = [].slice.call(loadingOverlays)
    loadingOverlays.forEach((loadingOverlay) => {
      loadingOverlay.remove()
    })
  }

  const createLoadingScreen = () => {
    removeLoadingScreen()
    const loadingOverlay = document.createElement('div')
    loadingOverlay.classList.add('vcv-loading-overlay')
    loadingOverlay.innerHTML = `<div class='vcv-loading-overlay-inner'>
        <div class='vcv-loading-dots-container'>
          <div class='vcv-loading-dot vcv-loading-dot-1'></div>
          <div class='vcv-loading-dot vcv-loading-dot-2'></div>
        </div>
      </div>`
    const startBlank = iframeContent.querySelector('.vcv-start-blank-container')
    if (startBlank) {
      iframeContent.insertBefore(loadingOverlay, startBlank)
    } else {
      iframeContent.appendChild(loadingOverlay)
    }
  }

  const reloadLayout = ({ type, template, header, sidebar, footer }) => {
    if (type === 'reload') {
      createLoadingScreen()
      const iframe = window.document.getElementById('vcv-editor-iframe')
      const domContainer = iframe.contentDocument.getElementById('vcv-editor')
      if (domContainer) {
        ReactDOM.unmountComponentAtNode(domContainer)
      }
      iframe.onload = () => {
        assetsStorage.trigger('reset')
        const data = vcCake.getService('document').all()
        const visibleElements = Utils.getVisibleElements(data)

        const setIframeLoaded = () => {
          workspaceIFrame.set({ type: 'layoutLoaded' })
          elementsStorage.off('elementsRenderDone', setIframeLoaded)
        }

        elementsStorage.on('elementsRenderDone', setIframeLoaded)

        workspaceIFrame.set({ type: 'loaded' })
        window.setTimeout(() => {
          elementsStorage.trigger('updateAll', data)
          assetsStorage.trigger('updateAllElements', visibleElements)
        }, 1)
      }
      const url = iframe.src.split('?')
      let params = url[1].split('&')
      params = params.reduce((arr, item) => {
        let write = true
        if (item.indexOf('vcv-template') >= 0) {
          write = false
        }
        if (item.indexOf('vcv-nonce') >= 0) {
          write = false
        }
        if (vcCake.env('VCV_JS_THEME_LAYOUTS')) {
          if (
            item.indexOf('vcv-header') >= 0 ||
            item.indexOf('vcv-sidebar') >= 0 ||
            item.indexOf('vcv-footer') >= 0
          ) {
            write = false
          }
        }
        write && arr.push(item)
        return arr
      }, [])
      params.push(`vcv-nonce=${dataManager.get('pageEditableNonce')}`)
      if (template) {
        params.push(`vcv-template=${template.value}`)
        params.push(`vcv-template-type=${template.type}`)
        params.push(`vcv-template-stretched=${template.stretchedContent ? 1 : 0}`)

        let hasHeader = false
        let hasSidebar = false
        let hasFooter = false
        const layouts = dataManager.get('pageTemplatesLayouts')
        const currentLayoutType = layouts && layouts.find(item => item.type === template.type)
        if (currentLayoutType && currentLayoutType.values) {
          const currentTemplate = currentLayoutType.values.find(item => item.value === template.value)
          if (currentTemplate) {
            hasHeader = currentTemplate.header
            hasSidebar = currentTemplate.sidebar
            hasFooter = currentTemplate.footer
          }
        }
        const isLayoutTheme = template.type === 'vc-custom-layout' && (template.value + '').indexOf('theme:') !== -1
        if (template.type === 'theme' || isLayoutTheme) {
          hasHeader = true
          hasFooter = true
        }
        if (vcCake.env('VCV_JS_THEME_LAYOUTS')) {
          if (hasHeader && header) {
            params.push(`vcv-header=${header}`)
          }
          if (hasSidebar && sidebar) {
            params.push(`vcv-sidebar=${sidebar}`)
          }
          if (hasFooter && footer) {
            params.push(`vcv-footer=${footer}`)
          }
        }
      }
      url[1] = params.join('&')
      iframe.src = url.join('?')
    } else if (type === 'loaded') {
      renderLayout(true)
    } else if (type === 'layoutLoaded') {
      removeLoadingScreen()
    }
  }
  api.on('editor:mount', function () {
    removeLoadingScreen()
  })
  wordpressDataStorage.state('status').onChange((result) => {
    if (result && result.status && result.status === 'loadSuccess') {
      renderLayout()
    }
  })
})
