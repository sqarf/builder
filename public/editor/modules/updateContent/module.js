import vcCake from 'vc-cake'
import React from 'react'
import ReactDOM from 'react-dom'
import Editor from './lib/editor'
import { Provider } from 'react-redux'
import store from 'public/editor/stores/store'

vcCake.add('updateContent', (api) => {
  // disable this module on regular editor
  // TODO: check this functionality
  if (typeof window.vcvRebuildPostSave === 'undefined') {
    console.warn('TODO: check is vcvRebuildPostSave works properly')
    return
  }
  const iframe = document.getElementById('vcv-editor-iframe')
  const iframeWindow = iframe ? iframe.contentWindow : null
  const domContainer = iframeWindow ? iframeWindow.document.getElementById('vcv-editor') : null
  if (domContainer) {
    ReactDOM.render(
      <Provider store={store}>
        <Editor api={api} />
      </Provider>,
      domContainer
    )
  }
})
