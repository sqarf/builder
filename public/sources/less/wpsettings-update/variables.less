// variables
@vcv-transitionDuration: .2s;
@vcv-transitionTimingFunction: ease-in-out;

@vcv-status-table-border-color: #f1f1f1;

@vcv-status-color-valid: #6baa3c;
@vcv-status-color-invalid: #e97272;

@vcv-status-color-tooltip: #eee;
@vcv-status-background-color-tooltip: #333;
@vcv-dashboard-headers-color: #515162;
@vcv-dashboard-link-color: #586dc6;
@vcv-dashboard-link-color--hover: #515162;
@vcv-dashboard-toggle-bg-color-on: #283AA4;
@vcv-dashboard-toggle-bg-color-off: #5c5b5b;
@vcv-dashboard-save-btn-bg-color: #ffae27;
@vcv-dashboard-refresh-btn-bg-color: #586ec7;
@vcv-dashboard-refresh-btn-bg-color--hover: #515263;
@vcv-dashboard-text-color: #8E8E9E;
@vcv-dashboard-attribute-color: #5F5F70;

@vcv-dashboard-font-family: 'Roboto', sans-serif;
@vcv-dashboard-title-font-family: 'Montserrat', Arial, sans-serif;

@vcv-dashboard-text-font-size: 15px;
@vcv-dashboard-text-line-height: 24px;
@vcv-dashboard-h1-font-size: 32px;
@vcv-dashboard-h2-font-size: 18px;
@vcv-dashboard-button-font-size: 12px;

@vcv-dashboard-dropdown-sort-down-icon: "data:image/svg+xml,%3C%3Fxml version='1.0' encoding='UTF-8' standalone='no'%3F%3E%3Csvg width='12px' height='6px' viewBox='0 0 12 6' version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3E%3Cdefs%3E%3Cpolygon points='0,0 12,0 6,6' fill='%23586dc6' id='path-1' /%3E%3C/defs%3E%3Cuse xlink:href='%23path-1'%3E%3C/use%3E%3C/svg%3E";
