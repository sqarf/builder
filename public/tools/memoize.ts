export default function memoize<T>(func: Function): Function {
    const cache: Map<string, T> = new Map()
    return (...args: any[]): T => {
        const key: string = JSON.stringify(args)
        const cached: T | undefined = cache.get(key)
        if (cached) {
            return cached
        }
        const result: T = func(...args)
        cache.set(key, result)

        return result
    }
}
