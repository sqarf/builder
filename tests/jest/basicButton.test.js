/* global expect, it */
import vcCake from 'vc-cake'

import '../../public/variables'
import '../../public/editor/services/utils/service.js'
import '../../public/editor/services/dataManager/service.js'
import '../../public/editor/services/document/service.js'
import '../../public/editor/services/dataProcessor/service.js'
import '../../public/editor/services/hubElements/service.js'
import '../../public/editor/services/cook/service.js'
import '../../public/editor/services/modernAssetsStorage/service.js'
import '../../public/editor/services/api/service.js'
import '../../public/editor/services/wordpress-post-data/service'

import '../../public/config/wp-attributes'

// import '../../public/editor/stores/shortcodesAssets/storage'
import '../../public/editor/stores/elements/elementSettings'

import '../../public/editor/modules/elementLimit/module'

import '../../public/tools/updateHtmlWithServer'


import React from 'react'
import renderer from 'react-test-renderer'
// import './__mocks__/api'

// Register element first
import '../../elements/basicButton/basicButton/index'
import ButtonElement from '../../elements/basicButton/basicButton/component'


describe('Tests Basic Button element render', () => {
  test('Basic Button render', () => {
    const text = 'Hello jest from react'

    vcCake.env('platform', 'jest').start(()=>{})
    const atts = vcCake.getService('cook').get({tag:'basicButton', id: '123456789', buttonText: text}).toJS()
    const tree = renderer
      .create(<ButtonElement atts={atts} id='123456789' />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
