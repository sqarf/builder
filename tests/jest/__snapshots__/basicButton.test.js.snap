// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Basic Button element render Basic Button render 1`] = `
<div
  className="vce-button--style-basic-container vce-button--style-basic-container--align-left"
>
  <span
    className="vce-button--style-basic-wrapper vce"
    data-vce-do-apply="margin el-123456789"
    id="el-123456789"
  >
    <button
      className="vce-button vce-button--style-basic vce-button--style-basic--border-square vce-button--style-basic--size-medium vce-button--style-basic--color-557cbf--fff"
      data-vce-do-apply="padding border background  el-123456789"
    >
      Hello jest from react
    </button>
  </span>
</div>
`;
